Get started analyzer code fix

BillWagner opened the pull request:
Fixes #4836 
Fixes #6994 

Relies on dotnet/samples#220

@Thraka Do you want to look at this? It might be a fun way to learn the Roslyn APIs, if you are interested.

[Internal review link](https://review.docs.microsoft.com/dotnet/csharp/roslyn-sdk/tutorials/how-to-write-csharp-analyzer-code-fix?branch=pr-en-us-6824)

BillWagner committed:
bring in the original files from the roslyn wiki

No edits yet.

BillWagner committed:
stash - switching branches.

BillWagner committed:
sample updated

BillWagner committed:
add notes

BillWagner committed:
final proofread and ready for review

Thraka commented:
@BillWagner sure!

rpetrusha commented:
Closing and reopening to begin new build.

BillWagner committed:
fix build errors.

BillWagner committed:
fix links to source code.

BillWagner committed:
final set of build warnings.

Thraka commented on docs/csharp/roslyn-sdk/tutorials/how-to-write-csharp-analyzer-code-fix.md:
```diff
@@ -0,0 +1,615 @@
+
+Change `TestMethod1` to the following code:
+
+[!code-csharp[Don't report diagnostics on const declarations](~/samples/csharp/roslyn-sdk/Tutorials/MakeConst/MakeConst/MakeConst.Test/MakeConstUnitTests.cs#AlreadyConst "Test to ensure no diagnostics are reported on already const definition")]
```
Pretty much all of the linked code samples have really bad indentation and look bad in the article.

BillWagner replied:
I want to make sure I make the right fix for this comment.  The code in magenta are string declarations that represent the source code being analyzed in a test. This is a typical test format for a diagnostic. See [roslyn-analyzer tests](https://github.com/dotnet/roslyn-analyzers/blob/master/src/Microsoft.CodeAnalysis.Analyzers/UnitTests/FixAnalyzers/FixerWithFixAllAnalyzerTests.Fixer.cs) for examples. If that's what you're concerned about, I'll highlight that better when it is first introduced.

Or do you mean the other code in the includes?

Thraka replied:
No I mean in the preview article, all of the code samples pulled in are indented incorrectly.

BillWagner replied:
Thanks for walking in through this with me. 

I came up with a plan to make the sample tutorial more readable, and keep the unit test structure more clear. I'll update the samples this morning, and work on the article (addressing the other existing feedback) this afternoon.

Thraka reviewed (changes requested):
I haven't run through the tutorial yet. This feedback was just from reading and checking the markdown source. Next I'll actually follow the tutorial and create an analyzer/code fix

Thraka commented:
@BillWagner I used the installer linked in the article, which was successful. But when I started visual studio, it told me that the Roslyn SDK is incompatible and disabled itself. 

```
Microsoft Visual Studio Enterprise 2017 
Version 15.7.5
VisualStudio.15.Release/15.7.5+27703.2042
Microsoft .NET Framework
Version 4.7.03056
```

BillWagner commented:
@Thraka Regarding installation:

I'll fix it and push an update, but in the meantime, [the updated installation instructions](https://docs.microsoft.com/en-us/dotnet/csharp/roslyn-sdk/#installation-instructions) are already published.

pkulikov commented on docs/csharp/roslyn-sdk/tutorials/how-to-write-csharp-analyzer-code-fix.md:
````diff
@@ -0,0 +1,615 @@
+}
+```
+
+You'll see that they are reported as warnings and "light bulb" suggestions appear next to them when the editor caret is on the same line. Move the editor caret to one of the squiggly underlines and press Ctrl+. to display the suggestion. Notice that a preview window appears next to the suggestion menu showing what the code will look like after the Code Fix is invoked.
````
What about: `<kbd>Ctrl</kbd>+<kbd>.</kbd>` that renders as <kbd>Ctrl</kbd>+<kbd>.</kbd>? Does it work for docfx?

Thraka commented on docs/csharp/roslyn-sdk/tutorials/how-to-write-csharp-analyzer-code-fix.md:
```diff
@@ -0,0 +1,615 @@
+
+[!code-csharp[Register the node action](~/samples/csharp/roslyn-sdk/Tutorials/MakeConst/MakeConst/MakeConst/MakeConstAnalyzer.cs#RegisterNodeAction "Register a node action")]
+
+After that change, you can delete the `AnalyzeSymbol` method. This analyzer examines <xref:Microsoft.CodeAnalysis.CSharp.SyntaxKind.LocalDeclarationStatement?displayProperty=nameWithType>, not <xref:Microsoft.CodeAnalysis.SymbolKind.NamedType?displayProperty=nameWithType> statements. Then, put your cursor on the `AnalyzeNode` method, press **Ctrl+.** and generate the `AnalyzeNode` method.
```
Should you use keyboard shortcuts versus mouse instructions? The user is told to move the cursor, not really indicating keyboard or mouse input, but in general, cursor indicates mouse movement. So maybe a **right-click > Quick-actions and refactoring > Generate method** and then put in `(or press CTRL + .)`

OR just have them copy/paste the method skeleton to replace the `AnalyzeSymbol` method.

--EDIT--
OK after the next paragraph, I see that you're trying to show them that pasting in the code causes that pasted code to be evaluated and the existing analyzer saw that there was no method named `AnalyzeNode`. Maybe these details should be right after the instruction to paste instead of the after the instruction to delete the old method?

Thraka commented on docs/csharp/roslyn-sdk/tutorials/how-to-write-csharp-analyzer-code-fix.md:
```diff
@@ -0,0 +1,615 @@
+
+This change highlights an important design consideration for every analyzer you write: Visual Studio calls analyzers as the user is editing code. Register actions for the types of code changes that could create something your analyzer should examine. That helps minimize any performance impact for Visual Studio users that load your analyzer.
+
+Next, you should update the string resources:
```
I would think a lot of users do not necessarily use the string resources in visual studio, give some directions. Which project are these resources in? How do I get to them?

Thraka commented on docs/csharp/roslyn-sdk/tutorials/how-to-write-csharp-analyzer-code-fix.md:
```diff
@@ -0,0 +1,615 @@
+
+* Change `AnalyzerTitle` to "Variable can be made constant".
+* Change `AnalyzerMessageFormat` to "Can be made constant".
+* Change `Description` to "Make Constant".
```
`AnalyzerDescription` not `Description`

Thraka commented on docs/csharp/roslyn-sdk/tutorials/how-to-write-csharp-analyzer-code-fix.md:
```diff
@@ -0,0 +1,615 @@
+        public const string DiagnosticId = "MakeConstCS";
+        private const string Title = "Variable can be made constant";
+        private const string MessageFormat = "Can be made constant";
+        private const string Description = "Make Constant";
```
This code does not match the file that was generated. This is what I have after following the directions

```csharp
using System;
using System.Collections.Generic;
using System.Collections.Immutable;
using System.Linq;
using System.Threading;
using Microsoft.CodeAnalysis;
using Microsoft.CodeAnalysis.CSharp;
using Microsoft.CodeAnalysis.CSharp.Syntax;
using Microsoft.CodeAnalysis.Diagnostics;

namespace MakeConst
{
    [DiagnosticAnalyzer(LanguageNames.CSharp)]
    public class MakeConstAnalyzer : DiagnosticAnalyzer
    {
        public const string DiagnosticId = "MakeConst";

        // You can change these strings in the Resources.resx file. If you do not want your analyzer to be localize-able, you can use regular strings for Title and MessageFormat.
        // See https://github.com/dotnet/roslyn/blob/master/docs/analyzers/Localizing%20Analyzers.md for more on localization
        private static readonly LocalizableString Title = new LocalizableResourceString(nameof(Resources.AnalyzerTitle), Resources.ResourceManager, typeof(Resources));
        private static readonly LocalizableString MessageFormat = new LocalizableResourceString(nameof(Resources.AnalyzerMessageFormat), Resources.ResourceManager, typeof(Resources));
        private static readonly LocalizableString Description = new LocalizableResourceString(nameof(Resources.AnalyzerDescription), Resources.ResourceManager, typeof(Resources));
        private const string Category = "Usage";

        private static DiagnosticDescriptor Rule = new DiagnosticDescriptor(DiagnosticId, Title, MessageFormat, Category, DiagnosticSeverity.Warning, isEnabledByDefault: true, description: Description);

        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get { return ImmutableArray.Create(Rule); } }

        public override void Initialize(AnalysisContext context)
        {
            // TODO: Consider registering other actions that act on syntax instead of or in addition to symbols
            // See https://github.com/dotnet/roslyn/blob/master/docs/analyzers/Analyzer%20Actions%20Semantics.md for more information
            context.RegisterSyntaxNodeAction(AnalyzeNode, SyntaxKind.LocalDeclarationStatement);
        }

        private void AnalyzeNode(SyntaxNodeAnalysisContext obj)
        {
            throw new NotImplementedException();
        }
    }
}
```

Thraka commented on docs/csharp/roslyn-sdk/tutorials/how-to-write-csharp-analyzer-code-fix.md:
```diff
@@ -0,0 +1,615 @@
+
+![Update string resources](media/how-to-write-csharp-analyzer-code-fix/update-string-resources.png)
+
+Also, change the `Category` to "Usage" as shown in the following code:
```
This being the last file edit to **MakeConstAnalyzer.cs** for this sub section, I would move this instruction up before the `change string resources` so you don't pop them in/out of the code editor for a small change.

Thraka commented on docs/csharp/roslyn-sdk/tutorials/how-to-write-csharp-analyzer-code-fix.md:
````diff
@@ -0,0 +1,615 @@
+Start by performing some syntactic analyzer when your action is invoked. Open **MakeConst.cs**. In the `AnalyzeNode` method, cast the node passed in to the <xref:Microsoft.CodeAnalysis.CSharp.Syntax.LocalDeclarationStatementSyntax> type. You can safely assume this cast will succeed because the registered action specifies that type of node.
+
+```C#
+var localDeclaration = (LocalDeclarationStatementSyntax)context.Node;
````
The `generate missing method` actually named my parameter as `obj` not `context`.

Thraka commented on docs/csharp/roslyn-sdk/tutorials/how-to-write-csharp-analyzer-code-fix.md:
```diff
@@ -0,0 +1,615 @@
+
+[!code-csharp[Don't report diagnostics on const declarations](~/samples/csharp/roslyn-sdk/Tutorials/MakeConst/MakeConst/MakeConst.Test/MakeConstUnitTests.cs#AlreadyConst "Test to ensure no diagnostics are reported on already const definition")]
+
+Run all the unit tests now. This new test passes. After all, your analyzer doesn't do anything yet. Add the following code to `AnalyzeNode` to ensure this test continues to pass:
```
How do I run unit tests?

Thraka commented on docs/csharp/roslyn-sdk/tutorials/how-to-write-csharp-analyzer-code-fix.md:
```diff
@@ -0,0 +1,615 @@
+
+[!code-csharp[Don't report diagnostics on multiple declarations if all can't be const](~/samples/csharp/roslyn-sdk/Tutorials/MakeConst/MakeConst/MakeConst.Test/MakeConstUnitTests.cs#MultipleDeclarations "Test to ensure no diagnostics are reported when multiple declarations can't all be made const")]
+
+As before, these tests will pass now. It's time to write the code that checks for these conditions. You'll perform some semantic analysis using the <xref:Microsoft.CodeAnalysis.Diagnostics.SyntaxNodeAnalysisContext>. You use the `context` argument to determine whether the local variable declaration can be made `const`. A <xref:Microsoft.CodeAnalysis.SemanticModel?displayProperty=nameWithType> rep ...
```
1. You wrote `you can earn more...` but you meant `you can learn more...`
2. Again watch out for that `context` variable as mine was named `obj`

Thraka commented on docs/csharp/roslyn-sdk/tutorials/how-to-write-csharp-analyzer-code-fix.md:
```diff
@@ -0,0 +1,615 @@
+
+### Create and report the diagnostic
+
+You're almost done. You've written the code to find a local declaration, and filtered out cases where a local declaration cannot be const. Now, let's write a test the shows a declaration that should be const. Open **MakeConstUnitTests.cs** and change `TestMethod2` to the following test:
```
Would it be easier to have them delete `TestMethod1` and `TestMethod2` at the start?

svick replied:
Typo: a test the shows → a test that shows

Thraka reviewed (changes requested):
These comments are from me running through the article and trying it out. It only covers the `Create the analyzer` section.

BillWagner commented on docs/csharp/roslyn-sdk/tutorials/how-to-write-csharp-analyzer-code-fix.md:
```diff
@@ -0,0 +1,615 @@
+
+* In Visual Studio, choose **File -> New -> Project...** to display the New Project dialog.
+* Under **Visual C# -> Extensibility**, choose **Analyzer with Code Fix (NuGet + VSIX)**.
+* Name your project "**MakeConst**" and click OK.
```
The default works fine.

BillWagner commented on docs/csharp/roslyn-sdk/tutorials/how-to-write-csharp-analyzer-code-fix.md:
```diff
@@ -0,0 +1,615 @@
+> [!TIP]
+> When you run your analyzer, you start a second copy of Visual Studio. This second copy uses a different registry hive to store settings. That enables you to differentiate the visual settings in the two copies of Visual Studio. You can pick a different theme for the experimental run of Visual Studio. In addition, don't roam your settings or login to your Visual Studio account using the experimen ...
+  
+In the second Visual Studio instance that you just started, create a new C# Console Application project. Hover over the token with a wavy underline, and the warning text provided by an Analyzer appears.
```
Either works. (added explanatory text).

svick commented on docs/csharp/roslyn-sdk/tutorials/how-to-write-csharp-analyzer-code-fix.md:
````diff
@@ -0,0 +1,615 @@
+        return;
+    }
+}
+```
````
If you have a declaration like `int i = 42, j;`, this will check `GetConstantValue().HasValue` on the `42` expression, which is an expensive semantic check, even though the cheap syntactic check for whether `j` has initializer will be `false`. This could be solved by splitting the loop into two: one checks `Initializer == null`, the other `GetConstantValue().HasValue`.

Is that kind of performance optimization appropriate here, or is simplicity more important? (In which case, this code would be fine.)

BillWagner replied:
I think keeping the explanation more clear works best here. Especially because the vast majority of variable declarations declare one variable.

svick commented on docs/csharp/roslyn-sdk/tutorials/how-to-write-csharp-analyzer-code-fix.md:
```diff
@@ -0,0 +1,615 @@
+
+### Make sure the variable isn't assigned after initialization
+
+You'll use the <xref:Microsoft.CodeAnalysis.SemanticModel?displayProperty=nameWithType> to perform data flow analysis on the local declaration statement. Then, you use the results of this data flow analysis to ensure that none of the local variables are written with a new value anywhere else. Call the <xref:Microsoft.CodeAnalysis.ModelExtensions.GetDeclaredSymbol%2A> extension method to retrieve  ...
```
> Call the \<xref:Microsoft.CodeAnalysis.ModelExtensions.GetDeclaredSymbol%2A> extension method

As far as I can tell, the code actually calls [the `CSharpExtensions.GetDeclaredSymbol` extension method](https://docs.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.csharp.csharpextensions.getdeclaredsymbol?view=roslyn-dotnet#Microsoft_CodeAnalysis_CSharp_CSharpExtensions_GetDeclaredSymbol_Microsoft_CodeAnalysis_SemanticModel_Microsoft_CodeAnalysis_CSharp_Syntax_VariableDeclaratorSyntax_System_Threading_CancellationToken_).

svick replied:
> contained with the `DataFlowAnalysis.WrittenOutside` collection

Should this be "contained with<ins>in</ins> the `DataFlowAnalysis.WrittenOutside` collection?

svick commented on docs/csharp/roslyn-sdk/tutorials/how-to-write-csharp-analyzer-code-fix.md:
```diff
@@ -0,0 +1,615 @@
+
+Add the following code to the end of the `MakeConstAsync` method:
+
+[!code-csharp[Format the new declaration](~/samples/csharp/roslyn-sdk/Tutorials/MakeConst/MakeConst/MakeConst/MakeConstCodeFixProvider.cs#ReplaceDocument  "Format the new declaration")]
```
Another snippet with wrong descriptions.

svick commented on docs/csharp/roslyn-sdk/tutorials/how-to-write-csharp-analyzer-code-fix.md:
```diff
@@ -0,0 +1,615 @@
+
+### Improve the quality
+
+Your analyzer works, but some C# expressions that aren't handled correctly:
```
Extra "that".

svick commented on docs/csharp/roslyn-sdk/tutorials/how-to-write-csharp-analyzer-code-fix.md:
```diff
@@ -0,0 +1,615 @@
+* If a variable is declared with the `var` keyword, the Code Fix does the wrong thing and generates a `const var` declaration, which is not supported by the C# language. To fix this bug, the code fix must replace the `var` keyword with the inferred type's name.
+Fortunately, all of the above bugs can be addressed using the same techniques that you just learned.
+
+[!code-csharp[Var declarations require types when constant](~/samples/csharp/roslyn-sdk/Tutorials/MakeConst/MakeConst/MakeConst.Test/MakeConstUnitTests.cs#ReplaceVarDeclarationWithConst "Var must be replaced with the type when made const")]
```
I think this snippet and the two above won't be formatted properly as part of the list and should be indented.

svick commented on docs/csharp/roslyn-sdk/tutorials/how-to-write-csharp-analyzer-code-fix.md:
````diff
@@ -0,0 +1,615 @@
+}
+```
+
+Fixing the third test requires a little more code to replace the var' keyword with the correct type name. Return to **CodeFixProvider.cs** and replace the code at the comment that reads "Produce the new local declaration" with the following code:
````
Stray `'` after "var".

svick commented on docs/csharp/roslyn-sdk/tutorials/how-to-write-csharp-analyzer-code-fix.md:
````diff
@@ -0,0 +1,615 @@
+                           .WithDeclaration(variableDeclaration);
+```
+
+Next, add a check inside curly braces of the if-block you wrote above to ensure that the type of the variable declaration is not an alias. If it is an alias to some other type (e.g. ), then it is legal to declare a local `const var`.
````
Missing example after "e.g.".

svick commented on docs/csharp/roslyn-sdk/tutorials/how-to-write-csharp-analyzer-code-fix.md:
```diff
@@ -0,0 +1,615 @@
+// to keep any leading and trailing trivia from the var keyword.
+var typeName = SyntaxFactory.ParseTypeName(type.ToDisplayString())
+    .WithLeadingTrivia(variableTypeName.GetLeadingTrivia())
+    .WithTrailingTrivia(variableTypeName.GetTrailingTrivia());
```
Wouldn't it be simpler to use `.WithTriviaFrom(variableTypeName)` here?

BillWagner committed:
major rewrite

This has some significant re-writes based on the earlier reviews. It should flow better and keep the user focused on the steps to update the analyzer.

Tests are used as good examples of code that should or should not raise the diagnostic.

BillWagner commented:
The latest changes rely on dotnet/samples#242

It's a rather sizable update, so a clean review makes the most sense.

/cc @thraka

BillWagner committed:
grammar check and proofread

BillWagner commented:
closing and re-opening to start a new build.

BillWagner committed:
fix build warnings.

A few mis-placed links.

Thraka commented on docs/csharp/roslyn-sdk/tutorials/how-to-write-csharp-analyzer-code-fix.md:
```diff
@@ -0,0 +1,468 @@
+
+[!code-csharp[Register the new code fix](~/samples/csharp/roslyn-sdk/Tutorials/MakeConst/MakeConst/MakeConstCodeFixProvider.cs#RegisterCodeFix  "Register the new code fix")]
+
+You'll notice red squiggles in the code you just added on the symbol `MakeConstAsync`. Right click on `MakeConstAsync` and select **Generate Method**. Your new `MakeConstAsync` method will transform the <xref:Microsoft.CodeAnalysis.Document> representing the user's source file into a new <xref:Microsoft.CodeAnalysis.Document> that now contains a `const` declaration.
```
Another place with the right-click steps.

Thraka commented on docs/csharp/roslyn-sdk/tutorials/how-to-write-csharp-analyzer-code-fix.md:
```diff
@@ -0,0 +1,468 @@
+
+You create a new `const` keyword token to insert at the front of the declaration statement. Be careful to first remove any leading trivia from the first token of the declaration statement and attach it to the `const` token. Add the following code to the `MakeConstAsync` method:
+
+[!code-csharp[Create a new const keyword token](~/samples/csharp/roslyn-sdk/Tutorials/MakeConst/MakeConst/MakeConstCodeFixProvider.cs#CreateConstToken  "Create the new const keyword token")]
```
My code gen created `declaration` not `localDeclaration`

BillWagner replied:
fixed by generating using copy / paste.

Thraka commented on docs/csharp/roslyn-sdk/tutorials/how-to-write-csharp-analyzer-code-fix.md:
```diff
@@ -0,0 +1,468 @@
+
+[!code-csharp[Format the new declaration](~/samples/csharp/roslyn-sdk/Tutorials/MakeConst/MakeConst/MakeConstCodeFixProvider.cs#FormatLocal  "Format the new declaration")]
+
+A new namespace is required for this code. Click the light bulb and add the suggested `using`. The final step is to make your edit. There are three steps to this process:
```
nit: This feels risky. What if the user disabled code suggestions? Or for whatever reason Visual Studio is not showing the lightbulb? Could ask them to just add `using Microsoft.CodeAnalysis.Formatting;`.

Thraka commented on docs/csharp/roslyn-sdk/tutorials/how-to-write-csharp-analyzer-code-fix.md:
```diff
@@ -0,0 +1,468 @@
+
+Next, format the new declaration to match C# formatting rules. Formatting your changes to match existing code creates a better experience. Add the following statement immediately after the existing code:
+
+[!code-csharp[Format the new declaration](~/samples/csharp/roslyn-sdk/Tutorials/MakeConst/MakeConst/MakeConstCodeFixProvider.cs#FormatLocal  "Format the new declaration")]
```
This code references a `newLocal` variable that doesn't exist. At the end of this section, it will not compile. I noticed there is a `newLocal` declaration at the end of the whole article though. Is it supposed to be `trimmedLocal`?

Thraka commented on docs/csharp/roslyn-sdk/tutorials/how-to-write-csharp-analyzer-code-fix.md:
```diff
@@ -0,0 +1,468 @@
+
+Add the following code to the end of the `MakeConstAsync` method:
+
+[!code-csharp[Format the new declaration](~/samples/csharp/roslyn-sdk/Tutorials/MakeConst/MakeConst/MakeConstCodeFixProvider.cs#ReplaceDocument  "Format the new declaration")]
```
The `await` operator throws an error because the method was generated without the `async` modifier. I had to add that in but I didn't see it in the steps to do so.

Stopping here because I'll have to start over when you get these pieces fixed :smile:

BillWagner committed:
respond to feedback

I've updated up to the first run with the code fix.

Thraka commented on docs/csharp/roslyn-sdk/tutorials/how-to-write-csharp-analyzer-code-fix.md:
```diff
@@ -0,0 +1,505 @@
+
+## Explore the analyzer template
+
+Your analyzer reports to the user any local variable declarations that can be converted to local constants. For example, consider the following code:
```
I was just thinking about this part of the section. It talks about how my analyzer reports any local vars that can be converted to consts. This section is named explore the template, the template doesn't do this (yet) and the next thing mentioned is running VS and seeing how it tells the user the class has lowercase letters. I think these intro paragraphs and samples should be moved down to the **Create analyzer registrations**  section

Thraka commented on docs/csharp/roslyn-sdk/tutorials/how-to-write-csharp-analyzer-code-fix.md:
````diff
@@ -0,0 +1,505 @@
+}
+```
+
+The preceding code also made a couple changes to the code that builds the expected diagnostic result. It uses the public constants registered in the `MakeConst` analyzer. In addition, it uses two string constants for the input and fixed source:
````
You don't specify that the two strings should be pasted at the top of the test cs file.

Thraka commented on docs/csharp/roslyn-sdk/tutorials/how-to-write-csharp-analyzer-code-fix.md:
```diff
@@ -0,0 +1,505 @@
+
+## Create tests for valid declarations
+
+As a general rule, analyzers should exist as quickly as possible, doing minimal work. Visual Studio calls registered analyzers as the user edits code. Responsiveness is a key requirement.  There are a number of tests to add for declarations that should not raise your diagnostic. There are several test cases for code that should not raise your diagnostic. Your analyzer already handles one of those ...
```
It seems you've repeated the same thought twice

> There are a number of tests to add for declarations that should not raise your diagnostic. There are several test cases for code that should not raise your diagnostic.

BillWagner replied:
And you can tell exactly where I was interrupted while writing.... 🙄

Thraka commented on docs/csharp/roslyn-sdk/tutorials/how-to-write-csharp-analyzer-code-fix.md:
```diff
@@ -0,0 +1,505 @@
+
+## Create tests for valid declarations
+
+As a general rule, analyzers should exist as quickly as possible, doing minimal work. Visual Studio calls registered analyzers as the user edits code. Responsiveness is a key requirement.  There are a number of tests to add for declarations that should not raise your diagnostic. There are several test cases for code that should not raise your diagnostic. Your analyzer already handles one of those ...
```
Did you mean "As a general rule, analyzers should **exist** as quickly as possible" ? Or did you mean **exit**

BillWagner replied:
exit

Thraka commented on docs/csharp/roslyn-sdk/tutorials/how-to-write-csharp-analyzer-code-fix.md:
```diff
@@ -0,0 +1,505 @@
+
+   [!code-csharp[declarations where the initializer isn't const](~/samples/csharp/roslyn-sdk/Tutorials/MakeConst/MakeConst.Test/MakeConstUnitTests.cs#InitializerNotConstant "a declaration where the initializer is not a compile-time constant should not raise the diagnostic")]
+
+All of the preceding conditions are complicated a C# rule that allows multiple declarations as part of one statement. Consider the following test case string constant:
```
The first sentence didn't make sense to me.

BillWagner replied:
reworded.

Thraka commented on docs/csharp/roslyn-sdk/tutorials/how-to-write-csharp-analyzer-code-fix.md:
```diff
@@ -0,0 +1,505 @@
+
+That sounds like a lot of code. It's not. Add the following code immediately after the initialization of `newModifiers`:
+
+[!code-csharp[Replace Var designations](~/samples/csharp/roslyn-sdk/Tutorials/MakeConst/MakeConst/MakeConstCodeFixProvider.cs#ReplaceVar "Replace a var designation with the explicit type")]
```
I needed to import the following namespace

```csharp
using Microsoft.CodeAnalysis.Simplification;
```

Thraka replied:
This code created a conflict with the `newLocal` variable. It already exists and is slightly different. The steps on where to paste this code did not specify that the existing `newLocal` variable needed to be deleted.

Thraka reviewed (changes requested):
Just about finished! I was able to run through with only 1 real bug at the end. I love the new flow of this a lot better than the last one. Great job!

BillWagner commented on docs/csharp/roslyn-sdk/tutorials/how-to-write-csharp-analyzer-code-fix.md:
```diff
@@ -0,0 +1,505 @@
+
+![Can make const warnings](media/how-to-write-csharp-analyzer-code-fix/make-const-warning.png)
+
+Notice that if you type const before each variable, the warnings are automatically removed. Additionally, changing a variable to const can affect the reporting of other variables. Add the `const` modifier to both `i` and `j`, and you get a new warning on `k` because it can now be `const`.
```
Good catch. I explained that some features are missing at this point in the tutorial. Then, I re-run this at the end to show that they are fixed.

BillWagner commented on docs/csharp/roslyn-sdk/tutorials/how-to-write-csharp-analyzer-code-fix.md:
```diff
@@ -0,0 +1,505 @@
+
+Your analyzer and code fix work on a simple case of a single declaration that can be made const. There are numerous possible declaration statements where this implementation makes mistakes. You'll address these cases by working with the unit test library written by the template. It's much faster than repeatedly opening a second copy of Visual Studio.
+
+Open the **MakeConstUnitTests.cs** file in the unit test project. The template created two tests that follow the two common patterns for an analyzer and code fix unit test. `Test1` shows the pattern for a test that ensures the analyzer doesn't report a diagnostic when it shouldn't. `Test2` shows the pattern for reporting a diagnostic and running the code fix.
```
globally changed.

BillWagner committed:
respond to feedback

Thanks for the thorough review @Thraka  All the changes have been addressed.

BillWagner commented:
Thanks @Thraka 

All the changes are addressed.
